In [1]:
import numpy as np
import pandas as pd
import os 
import statsmodels.api as sm
from plotnine import *
np.random.seed(12345)

### Problem 1

In [2]:
#load the data
df = pd.read_excel('vote1.xls', header = None, names = ['state', 'district', 'democA', 'voteA', 'expendA', 'expendB', 'prtystrA',  
'lexpendA', 'lexpendB', 'shareA'])    
df

state  district  democA  voteA  expendA  expendB  prtystrA  lexpendA  \
0    "AL"         7       1     68   328.30     8.74        41  5.793916   
1    "AK"         1       0     62   626.38   402.48        60  6.439952   
2    "AZ"         2       1     73    99.61     3.07        55  4.601233   
3    "AZ"         3       0     69   319.69    26.28        64  5.767352   
4    "AR"         3       0     75   159.22    60.05        66  5.070293   
..    ...       ...     ...    ...      ...      ...       ...       ...   
168  "WV"         4       0     39    32.04   152.27        42  3.466954   
169  "WI"         3       1     32    22.63   359.80        53  3.119100   
170  "WI"         5       0     36   197.46  1278.53        36  5.285536   
171  "WI"         7       0     38   202.59   450.72        46  5.311189   
172  "WI"         8       1     30    14.42   227.82        47  2.668685   

     lexpendB  shareA  
0    2.167567   97.41  
1    5.997638   60.88  
2    1.120048   97.01  
3    3.268846   92.40  
4    4.095244   72.61  
..        ...     ...  
168  5.025662   17.38  
169  5.885551    5.92  
170  7.153463   13.38  
171  6.110837   31.01  
172  5.428569    5.95  

[173 rows x 10 columns]

In [3]:
#log
df['log_expendA'] = np.log(df['expendA'])
df['log_expendB'] = np.log(df['expendB'])
df

state  district  democA  voteA  expendA  expendB  prtystrA  lexpendA  \
0    "AL"         7       1     68   328.30     8.74        41  5.793916   
1    "AK"         1       0     62   626.38   402.48        60  6.439952   
2    "AZ"         2       1     73    99.61     3.07        55  4.601233   
3    "AZ"         3       0     69   319.69    26.28        64  5.767352   
4    "AR"         3       0     75   159.22    60.05        66  5.070293   
..    ...       ...     ...    ...      ...      ...       ...       ...   
168  "WV"         4       0     39    32.04   152.27        42  3.466954   
169  "WI"         3       1     32    22.63   359.80        53  3.119100   
170  "WI"         5       0     36   197.46  1278.53        36  5.285536   
171  "WI"         7       0     38   202.59   450.72        46  5.311189   
172  "WI"         8       1     30    14.42   227.82        47  2.668685   

     lexpendB  shareA  log_expendA  log_expendB  
0    2.167567   97.41     5.793928     2.167910  
1    5.997638   60.88     6.439957     5.997645  
2    1.120048   97.01     4.601263     1.121678  
3    3.268846   92.40     5.767352     3.268808  
4    4.095244   72.61     5.070287     4.095178  
..        ...     ...          ...          ...  
168  5.025662   17.38     3.466985     5.025655  
169  5.885551    5.92     3.119276     5.885548  
170  7.153463   13.38     5.285536     7.153466  
171  6.110837   31.01     5.311184     6.110846  
172  5.428569    5.95     2.668616     5.428556  

[173 rows x 12 columns]

In [4]:
df.describe()

district      democA       voteA      expendA      expendB  \
count  173.000000  173.000000  173.000000   173.000000   173.000000   
mean     8.838150    0.554913   50.502890   310.611098   305.088497   
std      8.768823    0.498418   16.784761   280.985739   306.278015   
min      1.000000    0.000000   16.000000     0.300000     0.930000   
25%      3.000000    0.000000   36.000000    81.630000    60.050000   
50%      6.000000    1.000000   50.000000   242.780000   221.530000   
75%     11.000000    1.000000   65.000000   457.410000   450.720000   
max     42.000000    1.000000   84.000000  1470.670000  1548.190000   

         prtystrA    lexpendA    lexpendB      shareA  log_expendA  \
count  173.000000  173.000000  173.000000  173.000000   173.000000   
mean    49.757225    5.025556    4.944369   51.076185     5.025450   
std      9.983650    1.601602    1.571143   33.483895     1.601955   
min     22.000000   -1.197328   -0.072571    0.090000    -1.203973   
25%     44.000000    4.402246    4.095244   18.870000     4.402197   
50%     50.000000    5.492164    5.400558   50.850000     5.492156   
75%     56.000000    6.125580    6.110837   84.260000     6.125580   
max     71.000000    7.293476    7.344844   99.500000     7.293473   

       log_expendB  
count   173.000000  
mean      4.944403  
std       1.571067  
min      -0.072571  
25%       4.095178  
50%       5.400558  
75%       6.110846  
max       7.344842

In [5]:
#model
m = sm.OLS.from_formula('voteA ~ log_expendA + log_expendB + prtystrA', data = df)
results = m.fit()
#print the results
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Fri, 15 Apr 2022   Prob (F-statistic):           1.80e-57
Time:                        17:39:57   Log-Likelihood:                -596.88
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      45.0879      3.927     11.482      0.000      37.336      52.840
log_expendA     6.0814      0.382     15.915      0.000       5.327       6.836
log_expendB    -6.6156      0.379    -17.461      0.000      -7.364      -5.868
prtystrA        0.1520      0.062      2.451      0.015       0.030       0.274
==============================================================================
Omnibus:                        8.908   Durbin-Watson:                   1.604
Prob(Omnibus):                  0.012   Jarque-Bera (JB):                8.841
Skew:                           0.493   Prob(JB):                       0.0120
Kurtosis:                       3.505   Cond. No.                         344.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### (i) What is the interpretation of b1?
The sum of votes for candidate A increases by 6.0814 if log of campain expedentures by A rises by 1 unit.

#### (ii) In terms of the parameters, state the null hypothesis that a 1% increase in A’s expenditures is offset by a 1% increase in B’s expenditures.
 H0: b1=-b2 (1% increase in A’s expenditures is offset by a 1% increase in B’s expenditures).

 
 H1: b1!=-b2 (1% increase in A’s expenditures is not offset by a 1% increase in B’s expenditures).

$\theta$ = b1+b2 = 0


#### (iii) Estimate the given model using the data in VOTE1.RAW and report the results in usual form. Do A’s expenditures affect the outcome? What about B’s expenditures? Can you use these results to test the hypothesis in part (ii)?

In [28]:
#model
m1 = sm.OLS.from_formula('voteA ~ log_expendA + log_expendB', data = df)
results1 = m1.fit()
#print the results
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.785
Model:                            OLS   Adj. R-squared:                  0.783
Method:                 Least Squares   F-statistic:                     310.6
Date:                Thu, 07 Apr 2022   Prob (F-statistic):           1.72e-57
Time:                        12:26:51   Log-Likelihood:                -599.91
No. Observations:                 173   AIC:                             1206.
Df Residuals:                     170   BIC:                             1215.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      52.0509      2.750     18.926      0.000      46.622      57.480
log_expendA     6.3400      0.373     17.015      0.000       5.604       7.076
log_expendB    -6.7570      0.380    -17.784      0.000      -7.507      -6.007
==============================================================================
Omnibus:                        6.666   Durbin-Watson:                   1.663
Prob(Omnibus):                  0.036   Jarque-Bera (JB):                6.306
Skew:                           0.426   Prob(JB):                       0.0427
Kurtosis:                       3.387   Cond. No.                         34.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

All variables are significant (p-value about 0).

* Intercept - number of votes for candidate A = 45.0879 votes (with all other variables = 0). 
* log_expendA - number of votes for candidate A will increase by 6.3400	 for every log of expedenture by A increases by 1 (with all other variables = 0).
* log_expendB - number of votes for candidate A will decrease by 6.7570 for every log of expedenture by A increases by 1 (with all other variables = 0).




#### (iv) Estimate a model that directly gives the t statistic for testing the hypothesis in part (ii). What do you conclude? (Use a two-sided alternative.)

The hypothesis of the part (ii) can be written as $\theta$ = b1+b2 = 0 or b1 = $\theta$ - b2
 

In [17]:
t_values_log_expendA = results.params[1]/results.bse[1]
t_values_log_expendA

15.915032281326853

In [18]:
t_values_log_expendB = results.params[2]/results.bse[2]
t_values_log_expendB

-17.460603479449773

In [20]:
from scipy.stats import t, pearsonr, f

In [47]:
print(not (t.cdf(t_values_log_expendA, len(df) - 4) > 0.975) or (t.cdf(t_values_log_expendA, len(df) - 4) < 0.025))
print(not (t.cdf(t_values_log_expendA, len(df) - 4) > 0.995) or (t.cdf(t_values_log_expendA, len(df) - 4) < 0.005))

True
True


The H0 is true.

#### Addition Use F-test to evaluate whether adding prtystrA to a model significantly improves the model fit.

In [34]:
F = ((results.rsquared - results1.rsquared)/2)/((1 - results.rsquared)/(len(df)-4))
print('p-value, F-test:', 1 - f.cdf(F, 2, len(df)-4))

p-value, F-test: 1.0


In [46]:
F = ((results1.rsquared - results.rsquared)/2)/((1 - results1.rsquared)/(len(df)-3))
print('p-value, F-test:', 1 - f.cdf(F, 2, len(df)-3))

p-value, F-test: 1.0


### Problem 2

y = 28 + 2X_1 − 5X_2 + 7X_3 +u 

I. If X1 = 2, X2 = 5 and X3 = 7, the conditional expectation of Y given X1, X2 and X3:
y = 28 + 2x2 − 5x5 + 7x7 + u = 

In [27]:
#if we assume that errors = 0
y1 = 28 + 2*2 - 5*5 + 7*7
y1

56

II. Suppose now X1 = 5:

In [28]:
y2 = 28 + 2*5 - 5*5 + 7*7
y2

62

III. Do Y from the first two parts have the same variance? Explain.

In [34]:
X1_1 = 2
X1_2 = 5
X2 = 5
X3 = 7

In [37]:
#we use estimated variances
mu1 = (X1_1 + X2 + X3)/3 #mean of the sample 1
mu2 = (X1_2 + X2 + X3)/3 #mean of the sample 2
n = 1000-1 
variance1 = ((X1_1 - mu1)**2 + (X2 - mu1)**2 + (X3 - mu1)**2)/n
variance2 = ((X1_2 - mu2)**2 + (X2 - mu2)**2 + (X3 - mu2)**2)/n
print(variance1,variance2)

0.012666666666666666 0.002666666666666667


Variance for y1 is bigger than for y2. The variances are not the same because in the y2 two values are the same (x1=x2=5), which means that the variance will be less.

IV. Use the loop to draw 500 samples of Y. 

In [40]:
X_1 = np.random.normal(2, 9, 1000)
X_2 = np.random.poisson(16, 1000)
X_3 = np.random.poisson(4, 1000)

In [38]:
σ=4**(1/2)

In [55]:
Y = []
for x in range(500):
    Y.append(28 + 2*X_1 + 5*X_2 - 7*X_3 + np.random.normal(0,2,1000)) 

V. For each drawn sample run the OLS regression

In [56]:
for x in range(500):
    Y = 28 + 2*X_1 -5*X_2 + 7*X_3 + np.random.normal(0,2,1000)
    data_estimation = pd.DataFrame({'Y': Y,
                                    'X_1': X_1,
                                    'X_2': X_2,
                                    'X_3': X_3})
    model = sm.OLS.from_formula('Y ~ X_1 + X_2 + X_3', data = data_estimation).fit()

VII. For each parameter, what is the mean across all samples? Are these means close to
actual population parameters? Discuss how this result is related to the properties of OLS
estimator.

In [86]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.996
Model:                            OLS   Adj. R-squared:                  0.996
Method:                 Least Squares   F-statistic:                 7.683e+04
Date:                Sun, 17 Apr 2022   Prob (F-statistic):               0.00
Time:                        00:23:48   Log-Likelihood:                -2087.0
No. Observations:                1000   AIC:                             4182.
Df Residuals:                     996   BIC:                             4202.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     28.0216      0.283     99.070      0.000      27.467      28.577
X_1            1.9901      0.007    291.733      0.000       1.977       2.004
X_2           -4.9841      0.015   -329.220      0.000      -5.014      -4.954
X_3            6.9684      0.032    219.718      0.000       6.906       7.031
==============================================================================
Omnibus:                        2.196   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.333   Jarque-Bera (JB):                2.269
Skew:                           0.103   Prob(JB):                        0.322
Kurtosis:                       2.892   Cond. No.                         79.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Actual population mean for x_1 = 9 while x_2 and x_2 don't have means as they are poisson distributed. 

VIII. For each parameter, what is the percentage of samples where confidence interval
“hit” the actual true population value? Is it close to 0.95?

to know this we need to see what confidens intervals do not cover 0:

In [126]:
df = pd.DataFrame(data=[conf_interval1[0],conf_interval2[0]]).transpose()
df

0          1
0     7.763396   7.918921
1     8.644137   8.844211
2    10.360292  10.666172
3     8.130023   8.303297
4    13.619633  14.197313
..         ...        ...
995   6.530996   6.635111
996   9.191584   9.422666
997   9.247175   9.481549
998   8.565760   8.761604
999   8.112008   8.284383

[1000 rows x 2 columns]

In [127]:
filter1 = df[0]>0  

In [128]:
filter2=df[1]>0

In [129]:
filter3 = df[0]<0 

In [130]:
filter4= df[1]<0

In [131]:
df1 = df.where(filter1 & filter2)

In [132]:
df2 = df.where(filter3 & filter4)

In [138]:
df1 = df1.dropna()

In [139]:
df2 = df2.dropna()

In [140]:
len(df1)

988

In [141]:
len(df2)

12

In [145]:
procent = (1000-(len(df1)+len(df2)))/1000
procent #all confidence intervals “hit” the actual true population value

0.0

### Ptoblem 3


Проверка на гетероскедастичность

In [16]:
m3 = sm.OLS.from_formula('voteA ~ prtystrA + democA + log_expendA + log_expendB', data = df)
results = m3.fit()
#print the results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     169.2
Date:                Fri, 15 Apr 2022   Prob (F-statistic):           8.27e-58
Time:                        18:23:58   Log-Likelihood:                -593.22
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      37.6679      4.737      7.952      

In [9]:
# obtain OLS residuals
df['resid_sq'] = results.resid**2
m4 = sm.OLS.from_formula('resid_sq ~ prtystrA + democA + log_expendA + log_expendB', data = df)
results = m4.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:               resid_sq   R-squared:                       0.072
Model:                            OLS   Adj. R-squared:                  0.050
Method:                 Least Squares   F-statistic:                     3.252
Date:                Fri, 15 Apr 2022   Prob (F-statistic):             0.0134
Time:                        17:44:03   Log-Likelihood:                -1980.7
No. Observations:                 173   AIC:                             3971.
Df Residuals:                     168   BIC:                             3987.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    1.936e+04   1.44e+04      1.344      

By regressing the residuals we test the correlation of errors and independent variables. Coefficients of Intercept, prtstystrA, democA, log_expendB are statistically significant, which means that we have heteroskedasticity in our sample.


ii Breusch-Pagan test for heteroskedasticity

In [13]:
import statsmodels.stats.api as sms
from statsmodels.compat import lzip

In [14]:
names = ['Lagrange multiplier statistic', 'p-value',
        'f-value', 'f p-value']

In [12]:
test = sms.het_breuschpagan(results.resid, results.model.exog)

In [15]:
lzip(names, test)

[('Lagrange multiplier statistic', 6.645556608948249),
 ('p-value', 0.155847084053431),
 ('f-value', 1.67782339855937),
 ('f p-value', 0.15740675734812615)]

H0: There is homoskedasticity.

p-value = 0.155847084053431, so we fail to reject the H0. We do not have sufficient evidence to say that heteroscedasticity is present in the regression model.
